pkg/controller/plan/scheduler/vsphere/scheduler.go

```diff
@@ -0,0 +1,117 @@
+package vsphere
+
+import (
+	"context"
+	liberr "github.com/konveyor/controller/pkg/error"
+	api "github.com/konveyor/forklift-controller/pkg/apis/forklift/v1alpha1"
+	"github.com/konveyor/forklift-controller/pkg/apis/forklift/v1alpha1/plan"
+	plancontext "github.com/konveyor/forklift-controller/pkg/controller/plan/context"
+	"github.com/konveyor/forklift-controller/pkg/controller/provider/web/vsphere"
+)
+
+//
+// Scheduler for migrations from ESX hosts.
+type Scheduler struct {
+	*plancontext.Context
+	MaxInFlight int
+	inflight    map[string]int
+	pending     map[string][]*pendingVM
+}
+
+type pendingVM struct {
+	status *plan.VMStatus
+	cost   int
+}
+
+//
+// Return the next VM to migrate.
+func (r *Scheduler) Next() (vm *plan.VMStatus, err error) {
+	err = r.buildSchedule()
+	if err != nil {
+		return
+	}
+	for _, vms := range r.schedulable() {
+		if len(vms) > 0 {
+			vm = vms[0].status
+		}
+	}
+	return
+}
+
+//
+// Determine how much host capcity is occupied
+// by running migrations across all plans for
+// the same provider, and determine which
+// VMs are still waiting to be started.
+func (r *Scheduler) buildSchedule() error {
+	r.inflight = make(map[string]int)
+	r.pending = make(map[string][]*pendingVM)
+
+	// Since we modify the plan VMStatuses in memory,
+	// we need to use the plan from the context rather
+	// than from the list of plans that are retrieved below.
+	for _, vmStatus := range r.Plan.Status.Migration.VMs {
+		obj, err := r.Source.Inventory.VM(&vmStatus.Ref)
+		if err != nil {
+			return err
+		}
+		vm := obj.(*vsphere.VM)
```

Comment:
Since the _ref_ has been resolved during validation, I would recommend using the inventory.Get(&vsphere.VM{}, vmStatus.Ref.ID) would be more consistent with other places in the code and the preferred usage of the client.